apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,330 @@
+use crate::{
+    agent::{
+        reporter_server::{Reporter, ReporterServer},
+        ReporterGraph, ReporterResponse,
+    },
+    report::Report,
+    ReporterStats, ReporterTrace, TracesAndStats,
+};
+use bytes::BytesMut;
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use prost::Message;
+use prost_types::Timestamp;
+use reqwest::Client;
+use std::collections::HashMap;
+use std::io::Write;
+use std::net::SocketAddr;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tokio::sync::mpsc::Sender;
+use tokio::sync::Mutex;
+use tokio::time::{interval, Duration, MissedTickBehavior};
+use tonic::transport::{Error, Server};
+use tonic::{Request, Response, Status};
+
+type QueryUsageMap = HashMap<String, TracesAndStats>;
+type GraphUsageMap = Arc<Mutex<HashMap<ReporterGraph, QueryUsageMap>>>;
+
+static DEFAULT_INGRESS: &str = "https://usage-reporting.api.apollographql.com/api/ingress/traces";
+static INGRESS_CLOCK_TICK: Duration = Duration::from_secs(5);
+static TRIGGER_BATCH_LIMIT: u32 = 50;
+
+/// Allows common transfer code to be more easily represented
+#[allow(clippy::large_enum_variant)]
+enum StatsOrTrace {
+    Stats(ReporterStats),
+    Trace(ReporterTrace),
+}
+
+impl StatsOrTrace {
+    fn get_traces_and_stats(&self) -> TracesAndStats {
+        match self {
+            StatsOrTrace::Stats(_) => TracesAndStats {
+                stats_with_context: vec![],
+                ..Default::default()
+            },
+            StatsOrTrace::Trace(_) => TracesAndStats {
+                trace: vec![],
+                ..Default::default()
+            },
+        }
+    }
+
+    fn graph(&self) -> Option<ReporterGraph> {
+        match self {
+            StatsOrTrace::Stats(s) => s.graph.clone(),
+            StatsOrTrace::Trace(t) => t.graph.clone(),
+        }
+    }
+
+    fn key(&self) -> String {
+        match self {
+            StatsOrTrace::Stats(s) => s.key.clone(),
+            StatsOrTrace::Trace(t) => t.key.clone(),
+        }
+    }
+}
+
+/// Accept Traces and Stats from clients and transfer to an Apollo Ingress
+pub struct ReportSpaceport {
+    addr: SocketAddr,
+    // This Map will only have a single entry if used internally from a router.
+    // (because a router can only be serving a single graph)
```

Comment:
The configuration approach supports that fairly easily. Just add an higher level "graphs" toml array.

It would be hard to support via Environment Variables I think.

Reply:
Just to clarify, I don't think we need to support different graph API keys, but I do think we would potentially want to support multiple schemas / multiple variants.  You're right that poses a challenge with just an environment variable.  (As a further consideration to the env convo, maybe it's actually just the API key that we may want to _certainly_ allow externalization of.)